internal/tofu/transform_provider.go

```diff
@@ -102,147 +103,96 @@ func (t *ProviderTransformer) Transform(g *Graph) error {
+							log.Printf("[TRACE] ProviderTransformer: %s uses inherited configuration %s", dag.VertexName(v), pp)
+							break
+						}
+						log.Printf("[TRACE] ProviderTransformer: looking for %s to serve %s", pp, dag.VertexName(v))
 					}
-					log.Printf("[TRACE] ProviderTransformer: looking for %s to serve %s", pp, dag.VertexName(v))
 				}
-			}
 
-			// If this provider doesn't need to be configured then we can just
-			// stub it out with an init-only provider node, which will just
-			// start up the provider and fetch its schema.
-			if _, exists := needConfigured[key]; target == nil && !exists {
-				stubAddr := addrs.AbsProviderConfig{
-					Module:   addrs.RootModule,
-					Provider: p.Provider,
-				}
-				stub := &NodeEvalableProvider{
-					&NodeAbstractProvider{
-						Addr: stubAddr,
-					},
+				if target == nil {
+					// TODO this error message is possibly misleading / may not even be able to be hit given config provider validation?
+					diags = diags.Append(tfdiags.Sourceless(
+						tfdiags.Error,
+						"Provider configuration not present",
+						fmt.Sprintf(
+							"To work with %s its original provider configuration at %s is required, but it has been removed. This occurs when a provider configuration is removed while objects created by that provider still exist in the state. Re-add the provider configuration to destroy %s, after which you can remove the provider configuration again.",
+							dag.VertexName(v), fullProviderAddr, dag.VertexName(v),
+						),
+					))
+					continue
 				}
-				m[stubAddr.String()] = stub
-				log.Printf("[TRACE] ProviderTransformer: creating init-only node for %s", stubAddr)
-				target = stub
-				g.Add(target)
-			}
 
-			if target == nil {
-				diags = diags.Append(tfdiags.Sourceless(
-					tfdiags.Error,
-					"Provider configuration not present",
-					fmt.Sprintf(
-						"To work with %s its original provider configuration at %s is required, but it has been removed. This occurs when a provider configuration is removed while objects created by that provider still exist in the state. Re-add the provider configuration to destroy %s, after which you can remove the provider configuration again.",
-						dag.VertexName(v), p, dag.VertexName(v),
-					),
-				))
-				break
-			}
-
-			// see if this is a proxy provider pointing to another concrete config
-			if p, ok := target.(*graphNodeProxyProvider); ok {
-				g.Remove(p)
-				target = p.Target()
-			}
+				// see if this is a proxy provider pointing to another concrete config
+				if p, ok := target.(*graphNodeProxyProvider); ok {
+					target = p.Target()
+				}
 
-			log.Printf("[DEBUG] ProviderTransformer: %q (%T) needs %s", dag.VertexName(v), v, dag.VertexName(target))
-			if pv, ok := v.(GraphNodeProviderConsumer); ok {
-				pv.SetProvider(target.ProviderAddr())
+				log.Printf("[DEBUG] ProviderTransformer: %q (%T) needs %s", dag.VertexName(v), v, dag.VertexName(target))
+				if pv, ok := v.(GraphNodeProviderConsumer); ok {
```

Comment:
This type check is redundant, we already know `v` is `GraphNodeProviderConsumer`.

Those lines could be moved out of switch-case:
```go
	pv.SetProvider(target.ProviderAddr())
	g.Connect(dag.BasicEdge(v, target))
```

Reply:
[1a5e1ec](https://github.com/opentofu/opentofu/pull/2069/commits/1a5e1ec2d8b9938424e22ce2ff24936c96a769d4)